py/replica_restart_policy_tests.py

```diff
@@ -0,0 +1,181 @@
+        api_client, self.namespace, self.name, "PS", ["Running"])
+      restart_time = tf_job_client.get_start_time_by_index(
+        api_client, self.namespace, self.name, "PS", 0)
+
+      logging.info("First start time: %s, restart time: %s",
+                   str(first_start_time), str(restart_time))
+
+      if restart_time <= first_start_time:
+        self.failure = "Job {0} in namespace {1} with restart policy OnFailure \
+          failed to restart the pod with exit_code 1".format(
+          self.name, self.namespace)
+        logging.error(self.failure)
+        return
+
+    elif replica_restart_policy == "Never":
+      tf_job_client.terminate_replicas(api_client, self.namespace, self.name,
+                                       "ps", 1, exit_code)
+      tf_job_client.wait_for_replica_type_in_phases(api_client, self.namespace,
+                                                    self.name, "PS", ["Failed"])
+
+    elif replica_restart_policy == "ExitCode" and exit_code == 1:
+      tf_job_client.terminate_replicas(api_client, self.namespace, self.name,
+                                       "ps", 1, exit_code)
+      tf_job_client.wait_for_replica_type_in_phases(api_client, self.namespace,
+                                                    self.name, "PS", ["Failed"])
+
+    else:
+      tf_job_client.wait_for_replica_type_in_phases(
+        api_client, self.namespace, self.name, "PS", ["Running"])
+      first_start_time = tf_job_client.get_start_time_by_index(
+        api_client, self.namespace, self.name, "PS", 0)
+
+      tf_job_client.terminate_replicas(api_client, self.namespace, self.name,
+                                       "ps", 1, exit_code)
+
+      tf_job_client.wait_for_replica_type_in_phases(
+        api_client, self.namespace, self.name, "PS", ["Running"])
+      restart_time = tf_job_client.get_start_time_by_index(
+        api_client, self.namespace, self.name, "PS", 0)
+
+      logging.info("First start time: %s, restart time: %s",
+                   str(first_start_time), str(restart_time))
+
+      if restart_time <= first_start_time:
+        self.failure = "Job {0} in namespace {1} with restart policy ExitCode \
+          failed to restart the pod with exit_code 130".format(
+          self.name, self.namespace)
+        logging.error(self.failure)
+        return
+
+    # Delete the TFJob.
+    tf_job_client.delete_tf_job(
+      api_client, self.namespace, self.name, version=self.tfjob_version)
+    logging.info("Waiting for job %s in namespaces %s to be deleted.",
+                 self.name, self.namespace)
+    tf_job_client.wait_for_delete(
+      api_client,
+      self.namespace,
+      self.name,
+      self.tfjob_version,
+      status_callback=tf_job_client.log_status)
+
+  # Verify that the pod is restarted even after the container exits with success.
+  # We terminate PS with exit_code=0, and verify it is restarted.
+  def test_restart_always(self):
```

Comment:
To completely cover the combinations, we should also test:
- Always restart, exit code 1 (restart)
- OnFailure, exit code 0 (no restart)
- Never, exit code 0 (no restart)